subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyMetadataRulesIntegrationTest.groovy

```diff
@@ -772,4 +772,145 @@ class DependencyMetadataRulesIntegrationTest extends AbstractModuleDependencyRes
         "dependencies"           | false
         "dependency constraints" | true
     }
+
+    def "a rule can provide a custom selection reason thanks to dependency reason"() {
```

Comment:
:x: I can't find similar coverage for reasons being available in the resolution result for:
1) A dependency declared in the current project with a reason provided
2) A transitive dependency declared in another project, with a reason provided
3) A transitive dependency declared in a published module with a reason provided.

I think we need general coverage for these cases. Hopefully we can do that by adding reasons to the existing test cases, and don't need to add new test cases just for 'reason'.

Reply:
Same for constraints with declared reasons.

Reply:
I have updated `ResolveTestFixture`, to make it possible to expect multiple reasons (that wasn't possible before, that's why I had added a test in `ResolutionResultApiIntegrationTest`. Then I have added some tests and updated existing ones to add coverage.